File: shared/src/types/level.ts
import type { Grid } from './grid';
import type { Item } from './item';

export type LevelTheme =
  | 'kitchen'
  | 'camping'
  | 'grocery'
  | 'holiday'
  | 'meal_prep';

export interface Constraint {
  id: string;
  description: string;
  points: number;
  type: 'zone' | 'adjacency' | 'count' | 'exclusion';
  params: Record<string, unknown>;
}

export interface Level {
  id: string;
  grid: Grid;
  items: Item[];
  constraints: Constraint[];
  theme: LevelTheme;
  difficulty: number;       // 0.0–1.0
  progressionIndex: number;
  optimalScore: number;     // server-computed
  isDaily: boolean;
  dailyDate: string | null;
}

File: backend/app/schemas/level.py
from datetime import date

from pydantic import BaseModel, Field


class CellSchema(BaseModel):
    row: int
    col: int
    zone: str = "standard"
    occupied: bool = False
    item_id: str | None = None


class GridSchema(BaseModel):
    rows: int = Field(ge=3, le=10)
    cols: int = Field(ge=3, le=10)
    cells: list[list[CellSchema]]


class ItemSchema(BaseModel):
    id: str
    name: str
    shape: list[list[int]]
    zone_requirement: str | None = None
    points: int
    color: str


class ConstraintSchema(BaseModel):
    id: str
    description: str
    points: int
    type: str
    params: dict = {}


class LevelGenerateRequest(BaseModel):
    difficulty: float = Field(default=0.3, ge=0.0, le=1.0)
    theme: str | None = None


class LevelResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    grid: dict
    items: list
    constraints: list
    theme: str
    difficulty: float
    progression_index: int
    optimal_score: int
    is_daily: bool
    daily_date: date | None
